src/core/lib/backoff/backoff.cc

```diff
@@ -18,63 +18,52 @@
+  const double zero_to_one_rand = rand() / (double)RAND_MAX;
+  return a + zero_to_one_rand * range;
 }
+}  // namespace
 
-/* Generate a random number between 0 and 1. */
-static double generate_uniform_random_number(uint32_t* rng_state) {
-  *rng_state = (1103515245 * *rng_state + 12345) % ((uint32_t)1 << 31);
-  return *rng_state / (double)((uint32_t)1 << 31);
+Backoff::Backoff(const Options& options) : options_(options) {
+  seed = (unsigned int)gpr_now(GPR_CLOCK_REALTIME).tv_nsec;
 }
 
-static double generate_uniform_random_number_between(uint32_t* rng_state,
-                                                     double a, double b) {
-  if (a == b) return a;
-  if (a > b) GPR_SWAP(double, a, b);  // make sure a < b
-  const double range = b - a;
-  return a + generate_uniform_random_number(rng_state) * range;
+Backoff::Result Backoff::Begin(grpc_exec_ctx* exec_ctx) {
+  current_backoff_ = options_.initial_backoff();
+  const grpc_millis initial_timeout =
+      std::max(options_.initial_backoff(), options_.min_connect_timeout());
+  const grpc_millis now = grpc_exec_ctx_now(exec_ctx);
+  return Backoff::Result{now + initial_timeout, now + current_backoff_};
 }
 
-grpc_backoff_result grpc_backoff_step(grpc_exec_ctx* exec_ctx,
-                                      grpc_backoff* backoff) {
-  backoff->current_backoff = (grpc_millis)(GPR_MIN(
-      backoff->current_backoff * backoff->multiplier, backoff->max_backoff));
+Backoff::Result Backoff::Step(grpc_exec_ctx* exec_ctx) {
+  current_backoff_ =
+      (grpc_millis)(std::min(current_backoff_ * options_.multiplier(),
+                             (double)options_.max_backoff()));
   const double jitter = generate_uniform_random_number_between(
-      &backoff->rng_state, -backoff->jitter * backoff->current_backoff,
-      backoff->jitter * backoff->current_backoff);
-  const grpc_millis current_timeout =
-      GPR_MAX((grpc_millis)(backoff->current_backoff + jitter),
-              backoff->min_connect_timeout);
-  const grpc_millis next_timeout = GPR_MIN(
-      (grpc_millis)(backoff->current_backoff + jitter), backoff->max_backoff);
+      -options_.jitter() * current_backoff_,
+      options_.jitter() * current_backoff_);
+  const grpc_millis current_timeout = std::max(
+      (grpc_millis)(current_backoff_ + jitter), options_.min_connect_timeout());
+  const grpc_millis next_timeout = std::min(
+      (grpc_millis)(current_backoff_ + jitter), options_.max_backoff());
   const grpc_millis now = grpc_exec_ctx_now(exec_ctx);
-  const grpc_backoff_result result = {now + current_timeout,
-                                      now + next_timeout};
-  return result;
+  return Backoff::Result{now + current_timeout, now + next_timeout};
 }
 
-void grpc_backoff_reset(grpc_backoff* backoff) {
-  backoff->current_backoff = backoff->initial_backoff;
-}
+void Backoff::Reset() { current_backoff_ = options_.initial_backoff(); }
+
+void Backoff::SetRandomSeed(uint32_t seed) { srand(seed); }
```

Comment:
A member function setting global state in a non-owned API seems a bit weird?

Reply:
Good point. Reverted back to our own RNG impl.